File: tags/milestone1/Noxa.Emulation.Psp.Bios.NativeHLE/KFile.cpp
// ----------------------------------------------------------------------------
// PSP Player Emulation Suite
// Copyright (C) 2006 Ben Vanik (noxa)
// Licensed under the LGPL - see License.txt in the project root for details
// ----------------------------------------------------------------------------

#include "Stdafx.h"
#include "KernelHelpers.h"
#include "KFile.h"
#include <malloc.h>
#include <string.h>

using namespace System;
using namespace Noxa::Emulation::Psp;
using namespace Noxa::Emulation::Psp::Bios;

#define STDBUFFERSIZE	2048

KFile::KFile( KDevice* device, IMediaItem^ item )
{
	Device = device;
	Item = item;

	bool readOnly = ( item->Attributes & MediaItemAttributes::ReadOnly ) == MediaItemAttributes::ReadOnly;
	CanWrite = !readOnly;
}

KFile::KFile( KDevice* device, IMediaItem^ item, Stream^ stream, bool readOnly )
{
	Device = device;
	Item = item;
	BoundStream = stream;
	IsOpen = true;
	CanWrite = !readOnly;
	CanSeek = true;
}

KFile::KFile( int specialFileHandle )
{
	IsStd = true;
	IsOpen = true;
	CanWrite = !( specialFileHandle == HSTDIN );
	CanSeek = false;
}

KFile::~KFile()
{
}

KStdFile::KStdFile( int specialFileHandle )
	: KFile( specialFileHandle == HSTDIN )
{
	StdType = specialFileHandle;
	UID = specialFileHandle;

	_buffer = ( byte* )malloc( STDBUFFERSIZE );
	_bufferIndex = 0;
}

KStdFile::~KStdFile()
{
	SAFEFREE( _buffer );
	_bufferIndex = 0;
}

void KStdFile::Write( const byte* buffer, int count )
{
	// Copy to buffer
	assert( _bufferIndex + count < STDBUFFERSIZE );
	if( _bufferIndex + count >= STDBUFFERSIZE )
		return;
	memcpy( _buffer + _bufferIndex, buffer, count );
	_bufferIndex += count;

	// Add trailing 0 - we assume input doesn't have it
	_buffer[ _bufferIndex ] = 0;
	_bufferIndex++;

	// If the buffer ends in \n, flush
	if( _buffer[ _bufferIndex - 2 ] == '\n' )
	{
		String^ temp = gcnew String( ( char* )_buffer, 0, _bufferIndex );
		Debug::WriteLine( String::Format( "{0}: {1}", ( StdType == HSTDERR ? "stderr" : "stdout" ), temp ) );
		_bufferIndex = 0;
	}
}

void KStdFile::Write( const char* string )
{
	int length = strlen( string );
	this->Write( ( const byte* )string, length );
}

void KStdFile::Write( String^ string )
{
	char* ptr = KernelHelpers::ToNativeString( string );
	if( ptr != NULL )
		this->Write( ( const byte* )ptr, string->Length );
	SAFEFREE( ptr );
}

File: Noxa.Emulation.Psp.Bios.ManagedHLE/Kernel Types/KFile.cs
// ----------------------------------------------------------------------------
// PSP Player Emulation Suite
// Copyright (C) 2006 Ben Vanik (noxa)
// Licensed under the LGPL - see License.txt in the project root for details
// ----------------------------------------------------------------------------

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Text;

using Noxa.Emulation.Psp.Media;
using System.IO;

namespace Noxa.Emulation.Psp.Bios.ManagedHLE
{
	enum KSpecialFileHandle
	{
		StdIn = 0,
		StdOut = 1,
		StdErr = 2,
	}

	class KFile : KHandle
	{
		public Kernel Kernel;

		public KDevice Device;
		public IMediaItem Item;

		public bool IsBlockAccess;
		public int FolderOffset;
		public Stream Stream;

		public bool IsOpen;
		public bool CanWrite;
		public bool CanSeek;

		public bool PendingClose;
		public long Result;

		public KFile( Kernel kernel, KDevice device, IMediaItem item )
		{
			Kernel = kernel;
			Device = device;
			Item = item;

			FolderOffset = 0;

			IsOpen = true;
			CanWrite = !( ( item.Attributes & MediaItemAttributes.ReadOnly ) == MediaItemAttributes.ReadOnly );
			CanSeek = false;
		}

		public KFile( Kernel kernel, KDevice device, IMediaItem item, Stream stream )
		{
			Kernel = kernel;
			Device = device;
			Item = item;

			FolderOffset = 0;

			Stream = stream;

			IsOpen = true;
			CanWrite = stream.CanWrite;
			CanSeek = stream.CanSeek;
		}

		public KFile( Kernel kernel )
		{
			Kernel = kernel;
			IsOpen = true;
			CanWrite = true;
			CanSeek = false;
		}
	}

	class KStdFile : KFile
	{
		private string _name;
		private StringBuilder _sb;

		public KStdFile( Kernel kernel, KSpecialFileHandle specialFileHandle )
			: base( kernel )
		{
			UID = ( uint )specialFileHandle;
			switch( specialFileHandle )
			{
				case KSpecialFileHandle.StdIn:
					_name = "stdin";
					break;
				case KSpecialFileHandle.StdOut:
					_name = "stdout";
					break;
				case KSpecialFileHandle.StdErr:
					_name = "stderr";
					break;
			}

			_sb = new StringBuilder();
		}

		public unsafe void Write( uint address, int length )
		{
			sbyte* p = ( sbyte* )Kernel.MemorySystem.Translate( address );
			string s = new string( p, 0, length );

			_sb.Append( s );
			if( _sb[ _sb.Length - 1 ] == '\n' )
			{
				string line = _sb.ToString().Trim();
				if( _name == "stdout" )
					Log.WriteLine( Verbosity.Normal, Feature.Stdout, line );
				else
					Log.WriteLine( Verbosity.Critical, Feature.Stdout, line );
				_sb.Remove( 0, _sb.Length );
			}
		}
	}
}
